Sources/GRPCHTTP2Core/Client/Connection/LoadBalancers/RoundRobinLoadBalancer.swift

```diff
@@ -0,0 +1,766 @@
+      // All subchannels are closed; finish the streams so the run loop exits.
+      self.event.continuation.finish()
+      self.input.continuation.finish()
+
+    case .none:
+      ()
+    }
+  }
+
+  private func handleSubchannelGoingAway(key: EndpointKey) {
+    switch self.state.withLockedValue({ $0.parkSubchannel(withKey: key) }) {
+    case .closeAndUpdateState(_, let connectivityState):
+      // No need to close the subchannel, it's already going away and will close itself.
+      if let connectivityState = connectivityState {
+        self.event.continuation.yield(.connectivityStateChanged(connectivityState))
+      }
+    case .none:
+      ()
+    }
+  }
+
+  private func handleCloseInput() {
+    switch self.state.withLockedValue({ $0.close() }) {
+    case .closeSubchannels(let subchannels):
+      // Publish a new shutdown state, this LB is no longer usable for new RPCs.
+      self.event.continuation.yield(.connectivityStateChanged(.shutdown))
+
+      // Close the subchannels.
+      for subchannel in subchannels {
+        subchannel.close()
+      }
+
+    case .closed:
+      // No subchannels to close.
+      self.event.continuation.yield(.connectivityStateChanged(.shutdown))
+      self.event.continuation.finish()
+      self.input.continuation.finish()
+
+    case .none:
+      ()
+    }
+  }
+}
+
+@available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
+extension RoundRobinLoadBalancer {
+  private enum State {
+    case active(Active)
+    case closing(Closing)
+    case closed
+
+    struct Active {
+      private(set) var aggregateConnectivityState: ConnectivityState
+      private var picker: Picker?
+
+      var endpoints: [Endpoint]
+      var subchannels: [EndpointKey: SubchannelState]
+      var parkedSubchannels: [EndpointKey: Subchannel]
+
+      init() {
+        self.endpoints = []
+        self.subchannels = [:]
+        self.parkedSubchannels = [:]
+        self.aggregateConnectivityState = .idle
+        self.picker = nil
+      }
+
+      mutating func updateConnectivityState(
+        _ state: ConnectivityState,
+        key: EndpointKey
+      ) -> OnSubchannelConnectivityStateUpdate {
+        if let changed = self.subchannels[key]?.updateState(state) {
+          guard changed else { return .none }
+
+          let subchannelToClose: Subchannel?
+
+          switch state {
+          case .ready:
+            if let index = self.subchannels.firstIndex(where: { $0.value.markedForRemoval }) {
+              let (key, subchannelState) = self.subchannels.remove(at: index)
+              self.parkedSubchannels[key] = subchannelState.subchannel
+              subchannelToClose = subchannelState.subchannel
+            } else {
+              subchannelToClose = nil
+            }
+
+          case .idle, .connecting, .transientFailure, .shutdown:
+            subchannelToClose = nil
+          }
+
+          self.refreshPicker()
+          let aggregateState = self.recomputeAggregateState()
```

Comment:
Given we're always calling `refreshPicker()` and `recomputeAggregateState()` as a pair, would it make sense to do both things inside the same function? I don't think we'd ever want to do one independently of the other, right?

Reply:
Actually, I see you've added `refreshPickerAndAggregateState()` but we are never calling it - I suppose it was just an oversight. Can we make the change?